rules/manager.go

```diff
@@ -520,9 +521,91 @@ func (m *Manager) Update(interval time.Duration, files []string) error {
 	wg.Wait()
 	m.groups = groups
 
+	ts := time.Now()
+	m.restoreForState(ts)
+
 	return nil
 }
 
+// restoreForState restores the 'for' state of the alerts
+// by looking up last ActiveAt from storage.
+func (m *Manager) restoreForState(ts time.Time) {
+
+	maxtMS := int64(model.TimeFromUnixNano(ts.UnixNano()))
+	for _, newg := range m.groups {
+		for _, rule := range newg.Rules() {
+			alertRule, ok := rule.(*AlertingRule)
+			if !ok {
+				continue
+			}
+			_, err := alertRule.Eval(m.opts.Context, ts, newg.opts.QueryFunc, newg.opts.ExternalURL)
```

Comment:
One simple solution that I can think of is first waiting [here](https://github.com/prometheus/prometheus/blob/4a4e8a7d3bab177aa446e8b0fd4de2c41bf5298d/cmd/prometheus/main.go#L560) till the db is open. And inside the `ruleManager.Run()`, delay for one scrape interval, and move the restoring part inside `run()` of every group. Does that sound right?

Reply:
That should do it offhand, just be careful that there's no restoration logic if a new group is added due to a config reload.

Reply:
Alerts could also depend on recording rules, so we really need to run those too. So we'll probably need a bit of a different control flow here

Reply:
Another way is:
1. Rule groups perform the first `Eval` as usual.
2. Then we perform restoration once all groups are done with the first `Eval` (a barrier).
3. After the restoration we continue the evaluations as usual.

Does this flow sound right? During restoration currently I am doing a dry `Eval` without persisting the samples, just to have the alerts updated in the rule. With this new flow the dry run can be skipped.

Reply:
Something like that sounds sane as a first pass. Even if an alert isn't firing from the first eval we can restore it, so it can potentially work on the 2nd (the 1st eval may not have enough data to work with to produce all the alerts it should).